Initial implementation. Provide catalog health check and surgery for four existing issues.

deiferni opened the pull request:
Initial implementation of catalog doctor. Can detect and fix known issues in plones `portal_catalog`. 
Current test-runs on two GEVER deployments and one onegov-ish deployment have been used to detect well known issues, all of which can now be fixed with the initial implementation in this PR.

_Reviewer advice: Implementation has been highly iterative as some of the issues were only discovered while implementing this. Many things have changed in between. I would recommend to have a look at the final files instead of the single commits as many of your comments on early commits may be in vain._

The initial implementation provides four different fixes:
- `RemoveExtraRid`: the object still exists and two rids have been generated for the same object, thus the object has duplicated entries in the catalog. (Most likely caused by combination of https://github.com/4teamwork/ftw.copymovepatches and catalog queries while handling an `IObjectMovedEvent` event.)
- `RemoveOrphanedRid`: the object no longer exists but there are orphaned entries left behind in the catalog.
- `ReindexMissingUUID`: restore inconsistent `UID` index entries
- `RemoveRidOrReindexObject`: explorative surgery, figure out if the object is still traversable and then reindex or remove the object.

Health issue detection and surgery for the detected issues are conceptually split into two separate steps. `CatalogHealthCheck` provides a health status for the catalog and lists unhealthy rids if it finds any. It can be run separately from the second step, the surgery provided by `CatalogDoctor`. Surgery attempts to fix known health issues. Surgery will only commit changes if a post-surgery healthcheck yields no more health issues.

Surgery and healthcheck are each available as a `zopectl.command`, `bin/instance doctor healthcheck` and `bin/instance doctor surgery`.


Loosely based on https://github.com/4teamwork/opengever.maintenance/tree/lg-find-catalog-inconsistencies. Changed the strategy of detecting issues in index/unindex. Mainly just comparing sets of index/unindex key/values does not detect all issues, i.e. it won't detect swapped key/value pairs. Example, the following situation:

index:
```
/foo ->  123
/bar -> 456
```
unindex:
```
123 -> /bar
456 -> /foo
```
would not be picked up. so i also added comparing the items. this is a bit theoretical as in practice we have not yet experienced that issue.

Once this PR is merged and the package is released the following scripts/fixes should be deprecated:
- https://github.com/4teamwork/opengever.maintenance/pull/113
- https://gist.github.com/maethu/a24bd1787391a1078c2a
- https://github.com/4teamwork/ftw.copymovepatches/blob/91aea70798b17c7670255313b528347477ed010f/ftw/copymovepatches/browser/catalog_fixes.py
- https://github.com/4teamwork/opengever.maintenance/commit/e33f5e85f55bb8520d5d582416a9403625b074ad

I have already opened issues for optional nice to haves that are not part of the initial implementation on purpose, see https://github.com/4teamwork/ftw.catalogdoctor/issues.

deiferni committed:
Add catalog checkup for rid/uid and data.

Check consistency of mapping from path to rid and vice versa, also check
catalog metadata for its contents.

deiferni committed:
Add method to report symptom to catalog checkup.

deiferni committed:
Add catalog index length health checks.

deiferni committed:
Split checkup result into its own class.

deiferni committed:
Improve testing for checkup, add more testcases.

deiferni committed:
Add command to run checkup via command line.

deiferni committed:
Allow attaching multiple paths per rid.

deiferni committed:
Drop unnecessary empty test.

deiferni committed:
Fix tests for plone 5.

deiferni committed:
Add quick subcommand description to readme.

deiferni committed:
Allow writing to a log.

deiferni committed:
Add doctor and surgery for extra rid.

deiferni committed:
Improve naming in various places.

- Split catalog module into surgery and checkup.
- Use unhealthy instead of abberation
- Use healthcheck instead of checkup
- shorten zopectl command name

deiferni committed:
Use set instead of dict to track symptoms.

deiferni committed:
Add surgery for orphaned rid.

The object that was used to catalog these entries is gone and cannot be found
at path any longer.

deiferni committed:
Provide consistent path sort order.

deiferni committed:
Provide more detailed surgery result.

deiferni committed:
Add commit and post-surgery healthcheck.

Only commit in surgery and only if the post-surgery health check is successful.
Also add '--dry-run` argument to prevent committing for a testrun.

deiferni committed:
Add UID index length to healthcheck.

The UID index contains one UID per plone object. The UID for each object is
unique, thus we know that for each object in the catalog there will also be one
entry in the UID index. Thus we can use this index as a sample to check the
catalog index health.

deiferni committed:
Add command tests, slightly rework arguments.

Default dryrun to true for checkup. Checkup must never cause persistent changes
and thus never commit.

deiferni committed:
Rename index_data to catalog_data.

We mostly check the catalogs internal data strucuture and not the indices. I
think not mentioning indices in the method name is better.

deiferni committed:
Consistently call index uuid_index.

This hopefully helps a bit to avoid confusion between the uuid_index, an index
in the catalog and the catalogs internal uid attribute which contains the
path->rid mapping.

deiferni committed:
Add healthcheck for uuid index data.

Check rids in index and unindex against each other and agains the catalogs
uid mapping (path->rid).
This seems to change some of our assumptions regarding how we should update
catalog index length.

deiferni committed:
Improve surgery index removal.

Remove manually from indices as unindex_object aborts too
early if the internal data is inconsistent.

deiferni committed:
Provide get_physical_path and get_rid methods.

Consistently provide these methods as helper methods on the base test class.

deiferni committed:
Report rids in catalog but absent from uuid index.

deiferni committed:
Add compat module for compatibility imports.

deiferni committed:
Add helper to make unhealty extra rid after move.

Creates an unhealthy catalog the same way as happens in production. Should then
be fixable by surgery.

Also document in what way the catalog is unhealthy and provide entry points for
further analysis while developing fixes.

deiferni committed:
Add debug helper to convert BTrees to python.

deiferni committed:
Add helper to make orphaned rid.

deiferni committed:
Add surgery to remove rid from UnIndex.

deiferni committed:
Add debug helper to pprint a btrees collection.

deiferni committed:
Add surgery for unindex_object call.

Some indices seem to be ablte to maintain their internal integrity thus we can
use the official `unindex_object` method to remove a rid from these indices.

deiferni committed:
Add utility function to find keys for a given rid.

deiferni committed:
Add surgery to remove rid from DateRangeIndex.

deiferni committed:
Add surgery to remove rid from BooleanIndex.

deiferni committed:
Add surgery to remove rid from UUIDIndex.

deiferni committed:
Pass a linked `Length` to decrease during removal.

Allow to pass a `Length` that has to be decreased once an index row is empty.
Slightly refactors coupling of `index._length` to the index which is most
reliably used to determine the correct length as this varies heavily between
different index implementations.

deiferni committed:
Add surgery to remove rid from ExtendedPathIndex.

deiferni committed:
Add surgery to reindex missing uuid.

deiferni committed:
Add helper to make missing UUID index entry.

deiferni committed:
Add test for extra rid removal surgery.

deiferni committed:
Add test for orphaned rid removal surgery.

deiferni committed:
Force process indexing queue before healthcheck.

deiferni committed:
Add helper do drop object from catalog indexes.

deiferni committed:
Add surgery to reindex obj or drop it.

deiferni committed:
Add plone 5.1 classifier.

deiferni committed:
Add surgery to drop extra rid not in indexes.

Re-use existing surgery for extra rid, that is not in the UUID index, but in
rid-->path mapping and catalog metadata.

deiferni committed:
Process indexing queue after fixes.

Some fixes reindex objects, to guarantee an up-to date catalog before
post-surgery healthcheck we process the indexing queue.

deiferni committed:
Also handle moved objects that are still present.

deiferni committed:
Remove orphaned rid not present in indexes.

lukasgraf commented:
Just a note for now:
The assertion that [looks for a redirect](https://github.com/4teamwork/ftw.catalogdoctor/pull/1/files#diff-2aea21352366b1815cc00c12f5f8c96dR369) doesn't work quite right yet. In AI we have a case of a document whose *containing dossier* has been moved, and therefore the redirect is registered for the dossier, not the document.

`restrictedTraverse` seems to handle this without a problem, probably because it iteratively traverses from the top down: At some point it will encounter the redirected dossier, resolve the redirect, and continues to build the path for the document relative to the new dossier location.

Possible solution: Broaden the check to test whether there's a registered redirect *for any object in the AQ chain*.

njohner commented on README.rst:
```diff
@@ -4,13 +4,50 @@
 
-Plone 4.3.x
+Lists inconsistencies detected in ``portal_catalog``. Finds inconsistencies by
+inspecting the catalogs internal data structures. It currently uses ``paths``
```
catalogs'internal

njohner commented on README.rst:
```diff
@@ -4,13 +4,50 @@
+catalog metadata to determine if the catalog is healthy or if there are
+problems. Healtcheck is a read-only operation and won't modify the catalog.
+
+It can be run as following:
```
as follows

njohner commented on README.rst:
```diff
@@ -4,13 +4,50 @@
+in production.
+
+
+It can be run as following:
```
follows

deiferni committed:
Fix for acquired objects moved into their parents.

Correct special case handling for objects that were moved into one of their
parent objects and thus can still be found via acquisition when their former
path is used as their id was not changed.
We assumed that the redirector is involved, this is not correct. Thus drop
redirector check and instead make sure that the acquisition chain of the wrapped
and unwrapped object is different.

njohner commented on ftw/catalogdoctor/exceptions.py:
```diff
@@ -0,0 +1,2 @@
+class CantPerformSurgery(Exception):
+    """Raised when a procedure can't not be performed."""
```
`can't not` -> `cannot`

njohner commented on ftw/catalogdoctor/command.py:
```diff
@@ -0,0 +1,152 @@
+    processQueue()
+
+    formatter.info('Performing post-surgery healthcheck:')
+    post_result = healthcheck_command(portal_catalog, args, formatter)
```
In the case of `there_is_nothing_we_can_do`, we probably don't need a `post-surgery healthcheck`, as we couldn't fix all issues in the first place...

deiferni replied:
will `return` in that case 👍

njohner commented on ftw/catalogdoctor/command.py:
```diff
@@ -0,0 +1,152 @@
+        formatter.info('The following unhealthy rids could not be fixed:')
+        for unhealthy_rid in there_is_nothing_we_can_do:
+            unhealthy_rid.write_result(formatter)
+            formatter.info('')
```
In the case of `there_is_nothing_we_can_do`, we probably don't need a `post-surgery healthcheck`, as we couldn't fix all issues in the first place... That means we could return here, no?

deiferni commented:
We have addressed the issue mentioned above in https://github.com/4teamwork/ftw.catalogdoctor/pull/1/commits/e4f94fbc986249c5420a894f9d4ababbb0258645, the assertion was not correct and worked just by coincidence as the redirector is not involved at all.

njohner commented on ftw/catalogdoctor/command.py:
```diff
@@ -0,0 +1,152 @@
+    parsed_args = parser.parse_args(args)
+    # healthceck never commits and thus we always enable the dryrun flag.
+    if parsed_args.command == 'healthcheck':
+        parsed_args.dryrun = True
```
This a bit confusing. If we always set `dryrun` to `True`, for the `healthcheck`, we could just as well not set it and just always do `transaction.doom()` in the `healthcheck`

deiferni replied:
have added a separate `_run_healthcheck` method that can be called from `surgery_command`. `healthcheck_command` now always dooms the transaction, the confusing block of code is gone.

deiferni commented:
for the record: we have just performed testruns on 4 "productive" deployments, it seems to be able to fix all known problems.

njohner commented on ftw/catalogdoctor/healthcheck.py:
```diff
@@ -0,0 +1,253 @@
+        self.uuid_index_index_length = uuid_index_index_length
+        self.uuid_index_unindex_length = uuid_index_unindex_length
+
+    def _make_unhealthy_rid(self, rid, path=None):
```
`_get_or_add_unhealthy_rid` would be a bit more accurate here

njohner commented on ftw/catalogdoctor/surgery.py:
```diff
@@ -0,0 +1,449 @@
+            'in_catalog_not_in_uuid_unindex',
+        ): RemoveRidOrReindexObject,
+    }
+
```
I'd either add a note in the `docstring` that symptoms always have to be sorted in the tuples used as keys, or make sure they are sorted:
```
surgeries = dict( (tuple(sorted(key)), value)  for key, value in surgeries.items() )
```

deiferni replied:
Have amended the docstring. I'd rather not do it automatically as i want them to be sorted in the code in any case. But sorting is enforced from the healtcheck: https://github.com/4teamwork/ftw.catalogdoctor/blob/e4f94fbc986249c5420a894f9d4ababbb0258645/ftw/catalogdoctor/healthcheck.py#L142.

njohner commented on ftw/catalogdoctor/surgery.py:
```diff
@@ -0,0 +1,449 @@
+
+        surgery = surgery_cls(self.catalog, self.unhealthy_rid)
+        surgery.perform()
+        return surgery
```
It's a bit surprising to me that this returns the surgery. I'd maybe have expected that this would perform the surgery and set a success attribute on the `CatalogDoctor` and maybe save the surgery as `self.surgery`.

But it's fine as is though. I realise you do not know after performing the surgery whether it succeeded or not.

deiferni replied:
My thinking was the surgery should somehow return a surgery result, or report. Yet a class seemed a bit overkill at the moment, so it returns the surgery that contains the result.

njohner commented on ftw/catalogdoctor/surgery.py:
```diff
@@ -0,0 +1,449 @@
+
+    def unindex_rid_from_all_catalog_indexes(self, rid):
+        for idx in self.catalog.indexes.values():
+            surgery = self.removal.get(type(idx))
```
Now you're just trying to confuse us! You have different `IndexSurgery` which you assign to the `surgery` attribute of a `Surgery`, but `IndexSurgery` is not a `Surgery` but some sort of sub-surgery!

I think it would be easier to read if you either dropped the analogy a bit or found two different terms here. So maybe keep `Surgery`, but replace `IndexSurgery` with `RepairIndex` for example. Or name the `IndexSurgery` base class as `RepairStep` or `SurgeryStep`. The specific subclasses of the `IndexSurgery` are fine.

Some other possibilities: 
* use `Treatment` for one of the two classes (`Surgery` or `IndexSurgery`)
* Use `Procedure` instead of `Surgery`
* Use `Intervention` instead of `Surgery`

deiferni replied:
I've almost seen that coming 😂. You're right `SurgeryStep` is a much better name here. Have also renamed some other variables to clarify their meaning.

njohner commented on ftw/catalogdoctor/surgery.py:
```diff
@@ -0,0 +1,449 @@
+                    linked_length.change(-1)
+
+    def _remove_rid_from_unindex(self, unindex):
+        """Remove all rids from a reverse index."""
```
`"""Remove rid from a reverse index."""`

njohner commented on ftw/catalogdoctor/surgery.py:
```diff
@@ -0,0 +1,449 @@
+    """Remove rid from a `BooleanIndex`.
+
+    Lazily skips inverting the boolean index which is deferred to the next
+    reindex operation by plone.
```
Comment as is, got me worried and had me look at the code of the `BooleanIndex`. How about :
```
Lazily skips checking whether the boolean index should be inverted or not, as this operation is 
functionally irrelevant and will happen during the next reindex operation by plone.
```
instead?

deiferni replied:
thanks, changed the comment to your suggestion 👍.

njohner commented on ftw/catalogdoctor/surgery.py:
```diff
@@ -0,0 +1,449 @@
+        self.unindex_rid_from_all_catalog_indexes(rid)
+        self.delete_rid_from_paths(rid)
+        self.delete_rid_from_metadata(rid)
+        self.change_catalog_length(-1)
```
This is the same treatment as in `RemoveExtraRid`, the only difference is two more conditions to verify first.
I think some code duplication could be avoided here, but we can do that in a future refactoring.

deiferni replied:
Could maybe put the pre-op checks into separate methods, to reduce duplication a bit. But i agree we can do that later.

njohner commented on ftw/catalogdoctor/surgery.py:
```diff
@@ -0,0 +1,449 @@
+
+    def get_surgery(self):
+        symptoms = self.unhealthy_rid.catalog_symptoms
+        return self.surgeries.get(symptoms, None)
```
I feel like you could complement the list of symptoms here with two more:
* `rid_maps_to_single_path = len(self.unhealthy_rid.paths) == 1`
* `object_is_traversable = portal.unrestrictedTraverse(self.unhealthy_rid.paths[0], None)`
Then your surgeries would not have to include more `if`s with different outcomes. This would also allow to refactor the surgeries and have them a bit more standardised and modular, with less code repetition. So maybe in the base class:
```
def perform(self):
    self.can_perform_surgery()
    self.perform_surgery()

def can_perform_surgery(self):
    raise NotImplementedError

def perform_surgery(self):
    raise NotImplementedError

def assert_path_in_catalog_uids(self):
    path = list(self.unhealthy_rid.paths)[0]
    if path not in self.catalog.uids:
        raise CantPerformSurgery(
            "Expected path to be present in catalog uids {}"
            .format(path))
```
And then subclasses that call specific checks and `IndexSurgery`s

```
def can_perform_surgery(self):
    self.assert_path_in_catalog_uids()
    ...
```

deiferni replied:
Similar to what i suggested above 😉. If you agree i'd like to defer to refactor this until we extend/add more functionality. Objections?

njohner replied:
No it's fine, I'd say we just make a ticket for that.

njohner commented on ftw/catalogdoctor/debug.py:
```diff
@@ -0,0 +1,58 @@
+
+    This method is intended to help displaying catalog data structures on the
+    command line for debugging. It can be uses in combination with pprint to
+    quickly analize the state of the catalogs internal data structures.
```
catalogs'internal data structures

njohner commented on ftw/catalogdoctor/debug.py:
```diff
@@ -0,0 +1,58 @@
+        return list(maybe_btrees)
+
+
+def pprint_btrees(btrees):
```
😍

njohner commented on ftw/catalogdoctor/tests/__init__.py:
```diff
@@ -1,17 +1,237 @@
+        return healthcheck.run()
+
+    def choose_next_rid(self):
+        """Return a rid for testing currently unused.
```
Return a currently unused rid for testing.

njohner commented on ftw/catalogdoctor/tests/__init__.py:
```diff
@@ -1,17 +1,237 @@
+    def choose_next_rid(self):
+        """Return a rid for testing currently unused.
+
+        It will avoid to return rids already used by the catalog or already
```
avoid returning

njohner commented on ftw/catalogdoctor/tests/__init__.py:
```diff
@@ -1,17 +1,237 @@
+        while (index in self.catalog.data
+               or index in self.catalog.paths
+               or index in self._chosen_rids):
+            index = randint(-2000000000, 2000000000)
```
We can get negative rids??

deiferni replied:
yes, no signed ints in python. and why not use the extra bit, right 😉.

njohner commented on ftw/catalogdoctor/tests/__init__.py:
```diff
@@ -1,17 +1,237 @@
+        """Return the catalog metadata for an object as dict.
+        """
+        self.maybe_process_indexing_queue()
+        rid = self.portal_catalog.getrid('/'.join(obj.getPhysicalPath()))
```
`rid = self.portal_catalog.getrid(self.get_physical_path())`

njohner commented on ftw/catalogdoctor/tests/__init__.py:
```diff
@@ -1,17 +1,237 @@
+        """Return the catalog index data for an object as dict.
+        """
+        self.maybe_process_indexing_queue()
+        rid = self.portal_catalog.getrid('/'.join(obj.getPhysicalPath()))
```
`rid = self.portal_catalog.getrid(self.get_physical_path())`

njohner replied:
or maybe `self.get_rid()`?

njohner commented on ftw/catalogdoctor/tests/__init__.py:
```diff
@@ -1,17 +1,237 @@
+        :param new_id: the target object id after the object has been moved.
+          An UUID4 is used if this argument is not supplied.
+
+        This simulates an isses that surfaces when a package causes a
```
typo: isses -> issue

njohner commented on ftw/catalogdoctor/tests/__init__.py:
```diff
@@ -1,17 +1,237 @@
+          An UUID4 is used if this argument is not supplied.
+
+        This simulates an isses that surfaces when a package causes a
+        premature reindex during ftw.copymovepatches optimized move.
```
Actually these 2 lines of comment seem superfluous, as it is repeated but more clearly below...

njohner commented on ftw/catalogdoctor/tests/__init__.py:
```diff
@@ -1,17 +1,237 @@
+
+        # move without firing events, disconnect ob from catalog
+        old_id = self.child.getId()
+        ob = self.parent._getOb(self.child.getId())
```
`ob = self.parent._getOb(old_id)`

njohner commented on ftw/catalogdoctor/tests/__init__.py:
```diff
@@ -1,17 +1,237 @@
+        :param obj: the object that will be deleted and leave an unhealthy rid
+                    behind.
+
+        This simulates an isses that surfaces when an object with an extra
```
isses -> issue

njohner commented on ftw/catalogdoctor/tests/__init__.py:
```diff
@@ -1,17 +1,237 @@
+        self.maybe_process_indexing_queue()
+        return ob
+
+    def rectalog_object_with_new_rid(self, obj):
```
rectalog -> recatalog

njohner commented on ftw/catalogdoctor/tests/__init__.py:
```diff
@@ -1,17 +1,237 @@
+    def rectalog_object_with_new_rid(self, obj):
+        """Make catalog unhealthy by recataloging an object with a new rid.
+
+        This will leave the old rid behind in in catalog metadata and in the
```
in in -> in

njohner commented on ftw/catalogdoctor/tests/test_command.py:
```diff
@@ -0,0 +1,190 @@
+            ' uid index index length: 1',
+            ' uid index unindex length: 1',
+            'Catalog data is unhealthy, found 1 unhealthy rids:',
+            'rid 98 (--no path--):',
```
how come this is always 98? AS I understand it you have a random rid chooser.

deiferni replied:
The catalog chooses the id by itself here, my random chooser is just used when we need new rids in tests to break the catalog.

The catalog just does https://github.com/zopefoundation/Products.ZCatalog/blob/9abf1fe6b3831d26dd9004d24c0bca816e1b64e3/src/Products/ZCatalog/Catalog.py#L298-L306

and i initialize it to `97` in 

https://github.com/4teamwork/ftw.catalogdoctor/blob/e4f94fbc986249c5420a894f9d4ababbb0258645/ftw/catalogdoctor/tests/__init__.py#L50

njohner replied:
Oh of course... stupid!

njohner commented on ftw/catalogdoctor/tests/test_surgery.py:
```diff
@@ -0,0 +1,271 @@
+            ),
+            result.get_symptoms(unhealthy_rid.rid))
+
+        surgery = RemoveExtraRid(self.catalog, unhealthy_rid)
```
Maybe check that the `CatalogDoctor` finds the expected surgery:
`self.assertEqual(RemoveExtraRid, CatalogDoctor(unhealthy_rid).get_surgery())`

njohner replied:
This comment is valid for all tests that follow

deiferni replied:
the tests predate the doctor, but it makes perfect sense to also test the lookup, amended 👍.

njohner commented on ftw/catalogdoctor/tests/test_surgery.py:
```diff
@@ -0,0 +1,271 @@
+            ),
+            unhealthy_rid.catalog_symptoms)
+
+        surgery = RemoveExtraRid(self.catalog, unhealthy_rid)
```
Same as above

njohner commented on ftw/catalogdoctor/tests/test_surgery.py:
```diff
@@ -0,0 +1,271 @@
+        self.make_unhealthy_extra_rid_after_move(self.child)
+
+        result = self.run_healthcheck()
+        self.assertFalse(result.is_healthy())
```
Add `self.assertEqual(1, len(result.get_unhealthy_rids()))`

njohner replied:
This comment is valid for all tests that follow

njohner commented on ftw/catalogdoctor/tests/test_surgery.py:
```diff
@@ -0,0 +1,271 @@
+
+    def test_surgery_add_dropped_object_to_indices(self):
+        self.drop_object_from_catalog_indexes(self.parent)
+
```
Can we assert here that there is no data in catalog? Something like

`self.assertEqual({}, self.get_catalog_indexdata(self.parent))`

njohner commented on ftw/catalogdoctor/tests/test_surgery.py:
```diff
@@ -0,0 +1,271 @@
+
+        result = self.run_healthcheck()
+        self.assertFalse(result.is_healthy())
+        unhealthy_rid = result.get_unhealthy_rids()[0]
```
Can we do a `self.assertEqual(rid, unhealthy_rid.rid)`

njohner commented on ftw/catalogdoctor/tests/test_catalog_healthcheck.py:
```diff
@@ -0,0 +1,266 @@
+        result = self.run_healthcheck()
+        self.assertFalse(result.is_length_healthy())
+
+    def test_exra_uid_index_make_catalog_unhealthy(self):
```
exra -> extra

njohner commented on ftw/catalogdoctor/tests/test_surgery_remove_from_boolean_index.py:
```diff
@@ -0,0 +1,95 @@
+        self.assertEqual(1, index._index_length.value)
+        self.assertEqual(1, len(index._unindex))
+        # off by one. not sure what is happening here. not observed
+        # in production. maybe test-setup related?
```
huh, that is pretty weird... I also looked at it and don't understand why

njohner commented on ftw/catalogdoctor/tests/test_surgery_remove_from_unindex.py:
```diff
@@ -0,0 +1,135 @@
+        self.grant('Contributor')
+        self.folder = create(Builder('folder').titled(u'Foo'))
+
+        # pretend we are something that supports recurring dates
```
I'd move that to the only test that needs it, namely `test_remove_healthy_object_from_daterecurringindex`

njohner reviewed (commented):
🎆 Looks good overall. Thanks for the `debug` thing, many comments and thorough testing.

Comments are mainly to improve code readability further and I think we could also avoid some code repetition. I did find the naming of `IndexSurgery` and `Surgery` quite confusing.

I think you pushed while I reviewed, so potentially I did not review the last commit or so

njohner commented:
Went through the last commit, LGTM

deiferni committed:
Skip post surgery healthcheck for unfixables.

deiferni committed:
Always doom transaction for healthcheck.

Explicitly doom transaction in healthcheck command, factor out healthcheck
helper function that can be called from surgery.

deiferni committed:
Improve unhealthy rid factory method name.

deiferni committed:
Indicate that symptom tuples shall be sorted.

deiferni committed:
Rename IndexSurgery to SurgeryStep.

Improves readability by avoiding to overload the name surgery.

deiferni committed:
Use helper methods to get rid instead of redefine.

deiferni committed:
Drop unused imports.

deiferni committed:
Test surgery lookup by doctor is correct.

Also improve testing and add another assertion here and there.

deiferni commented:
@njohner thanks for the great review! i went through all your comments, most of them are now implemented. i deferred the refactoring, unless you disagree we can do that at a later point in time. please have a look again, the non-typo changes went into separate commits at the end.

njohner reviewed (approved):
LGTM thanks!
Can you make an issue for the potential refactoring to limit code duplication?

deiferni commented:
Issue created in https://github.com/4teamwork/ftw.catalogdoctor/issues/5
